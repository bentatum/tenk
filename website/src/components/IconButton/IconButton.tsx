import { FC, forwardRef } from "react";
import classNames from "classnames";
import { BrandColors, StatusColors } from "@/types";
import { ButtonBase, ButtonBaseProps } from "../ButtonBase";

export interface IconButtonProps extends ButtonBaseProps {
  href?: string;
  color?: "neutral" | BrandColors | StatusColors;
  size?: "sm" | "md" | "lg";
  variant?: "text" | "contained";
}

export const IconButton: FC<IconButtonProps> = forwardRef(
  (
    {
      size = "md",
      variant = "text",
      color = "neutral",
      disabled = false,
      ...props
    },
    ref
  ) => {
    return (
      <ButtonBase
        {...props}
        ref={ref}
        className={classNames(
          "rounded-full p-1 min-h-full min-w-full",
          "uppercase inline-flex items-center justify-center",
          "transition ease-in duration-200 text-center",
          "focus:outline-none",
          "disabled:cursor-not-allowed disabled:bg-neutral-300 disabled:text-white",
          color === "neutral" && variant === "text"
            ? "hover:bg-gray-400/20 focus:ring-gray-500/50 dark:active:bg-gray-300/20 dark:hover:bg-gray-500/20 dark:focus:ring-gray-500/50 dark:active:bg-gray-300/20"
            : undefined,
          {
            "h-14 w-14": size === "lg",
            "h-10 w-10": size === "md",
            "h-9 w-9": size === "sm",
            "text-current": variant === "text",
            "drop-shadow focus:drop-shadow-none active:drop-shadow-none":
              variant === "contained",
            "bg-green-500 text-white focus:ring-pink-500/50":
              variant === "contained" && color === "green",
            "hover:bg-pink-500/20 focus:ring-pink-500/50":
              color === "green" && variant === "text",
            "text-teal-500 hover:bg-teal-500/50 focus:ring-teal-500":
              color === "teal",
          },
          props.className
        )}
      />
    );
  }
);

IconButton.displayName = "IconButton";

export default IconButton;
