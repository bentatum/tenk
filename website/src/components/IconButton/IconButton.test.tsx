import { render, screen } from "@testing-library/react";
import { SvgIcon } from "..";
import IconButton from "./IconButton";

const Icon = () => (
  <SvgIcon viewBox="0 0 24 24" className="w-6 h-6">
    <path
      strokeLinecap="round"
      strokeLinejoin="round"
      d="M6 18L18 6M6 6l12 12"
    />
  </SvgIcon>
);

describe("IconButton", () => {
  it("matches snapshot", () => {
    const { container } = render(
      <IconButton>
        <Icon />
      </IconButton>
    );
    expect(container).toMatchSnapshot();
  });

  describe("text variant", () => {
    describe("neutral color", () => {
      it("should be visible on hover and focus", () => {
        render(
          <IconButton color="neutral" variant="text">
            <Icon />
          </IconButton>
        );
        expect(screen.getByRole("button")).toHaveClass(
          "hover:bg-gray-400/20",
          "focus:ring-gray-500/50",
          "dark:active:bg-gray-300/20",
          "dark:hover:bg-gray-500/20",
          "dark:focus:ring-gray-500/50",
          "dark:active:bg-gray-300/20"
        );
      });
    });

    describe("pink color", () => {
      it("should have hover and focus colors", () => {
        render(
          <IconButton color="pink" variant="text">
            <Icon />
          </IconButton>
        );
        expect(screen.getByRole("button")).toHaveClass(
          "hover:bg-pink-500/20",
          "focus:ring-pink-500/50"
        );
      });
    });
  });

  describe("contained variant", () => {
    describe("neutral", () => {
      it("should not have the same hover and focus colors as text variant", () => {
        render(
          <IconButton color="neutral" variant="contained">
            <Icon />
          </IconButton>
        );
        expect(screen.getByRole("button")).not.toHaveClass(
          "hover:bg-gray-400/20",
          "focus:ring-gray-500/50",
          "dark:active:bg-gray-300/20",
          "dark:hover:bg-gray-500/20",
          "dark:focus:ring-gray-500/50",
          "dark:active:bg-gray-300/20"
        );
      });
    });
  });
});
