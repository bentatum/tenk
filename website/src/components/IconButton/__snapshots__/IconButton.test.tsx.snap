// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton matches snapshot 1`] = `
<div>
  <button
    class="focus:outline-none rounded-full p-1 min-h-full min-w-full uppercase inline-flex items-center justify-center transition ease-in duration-200 text-center focus:outline-none disabled:cursor-not-allowed disabled:bg-neutral-300 disabled:text-white hover:bg-gray-400/20 focus:ring-gray-500/50 dark:active:bg-gray-300/20 dark:hover:bg-gray-500/20 dark:focus:ring-gray-500/50 dark:active:bg-gray-300/20 h-10 w-10 text-current"
  >
    <svg
      aria-hidden="true"
      class="shrink-0 fill-current w-6 h-6"
      viewBox="0 0 24 24"
    >
      <path
        d="M6 18L18 6M6 6l12 12"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
  </button>
</div>
`;
